### ISSUE TITLE
Add (Implement ) DS Record for ClouDNS

### ISSUE THREAD

**Issue by jamaltebi** [StackExchange/dnscontrol#1000]

**Comment by tlimoncelli** [comment]
CC: @pragmaton

**Comment by jamaltebi** [comment]
@tlimoncelli. I think ClouDNS does not support yet adding DS record via API. I asked the help of ClouDNS

**Comment by jamaltebi** [comment]
@tlimoncelli  I changed the func `toReq()` ( to implement DS record), But I got this error: 

integration_test.go:188: failed create record (ClouDNS): ClouDNS API error: You can not add a DS record for the root domain. URL:api.cloudns.net/dns/add-record.json?auth-password=XXXXXX&domain-name=voitex.de&ds_Algorithm=13&ds_Digest=ADIGEST&ds_DigestType=1&ds_KeyTag=1&host=&record=&record-type=DS&sub-auth-id=XXXXX&ttl=300.

In ClouDNS webseite  https://www.cloudns.net/wiki/article/58/  there is no DS Record via API
![image](https://user-images.githubusercontent.com/51935048/103155196-8b73ed00-479d-11eb-865a-b58248b4ec67.png)

**Comment by tlimoncelli** [comment]
Ah, it looks like ClouDNS can do CanUseDSForChildren but not CanUseDS.  Give that a try.

**Comment by jamaltebi** [comment]
As you said, I found In ClouDNS , that we can add DS record just for Subdomain. I tried `CanUseDSForChildren `  I got this error

failed create record (ClouDNS): ClouDNS API error: You need to have NS records for this host to be able to add DS records for it. URL:api.cloudns.net/dns/add-record.json?

Even if  I add NS records for "child" ( child.mydomain) , I get the same error when I run the Test. Th test just delete the NS that I add for the child

**Comment by tlimoncelli** [comment]
Hmm... that's interesting! 

Are you adding the NS records using `NS()` or `NAMESERVER()`?  Can you share your dnsconfig.js file (remove anything that shouldn't be shared)

**Comment by jamaltebi** [comment]
Using `NS()` . I add NS record via Intarface. I run just the test for DS Child Record. The NS Record is included in this Test I see. It create and delete NS . when it come to DS Record , it fali because it must be a NS records ( from ClouDNS).

**Comment by tlimoncelli** [comment]
Is this an ordering problem?  i.e. Does the NS record have to be created before the DS record?  

Maybe ClouDNS support could help?

**Comment by jamaltebi** [comment]
I asked ClouDNS Help. Yes, the NS record must exist for the subdomain to wich I want add DS record:
"**Host**: host (You can not add a DS record for the root domain.) Please note that you need to have NS records for this **host** to be able to add DS records for it."  https://www.cloudns.net/wiki/article/365/

Fron Integration_test, I understand that, first it detect records that exist at providers ( in my case NS record for  child.mydomain). It delete this record and when trying to add DS record ( "create DS record child.mydomain ...." ) it find no NS record for child.mydomain(deleted when runing integration_test)

**Comment by tlimoncelli** [comment]
All records are removed from the zone at the beginning of each test. This is to "clean up" from past tests.

**Comment by jamaltebi** [comment]
In my case. I have : 
- NS reocd for mydomain
- NS record for host.mydomain
when I run Test for DS record. It delete  just NS record for host.mydomain( that DS record need ti be add)

But what I did when runing Test, after clean up. I add NS recod for my subdomain through Interface. The problem now is I get error : " invalid field Key Tag" even I used a valid DS record(I can add the same DS record via through Interface)

**Comment by tlimoncelli** [comment]
Looks like you are doing everything right.  I don't know what to suggest other than contacting ClouDNS support.

**Comment by jamaltebi** [comment]
Thanks. These are my Code-changes for Implement DS record:
https://github.com/jamaltebi/dnscontrol/commit/df65820bae5977ccba2a49d96fc19f6217488351
I asked CLouDNS many times via LiveChat, but they dont know why... I am wokring on why I get a Invalid Key Tag

**Comment by jamaltebi** [comment]
@tlimoncelli  I can add DS record If I use in the field **record** instead **digest**  .

like : 

```
case "DS":
		req["key-tag"] = strconv.Itoa(int(rc.DsKeyTag))
		req["algorithm"] = strconv.Itoa(int(rc.DsAlgorithm))
		req["digest-type"] = strconv.Itoa(int(rc.DsDigestType))
		//req["digest"] = rc.DsDigest
		req["record"] = rc.DsDigest
```

but the integration test fails.
I don't understand good this part in func `toReq()`

```
req := requestParams{
		"record-type": rc.Type,
		"host":        rc.GetLabel(),
		"record":      rc.GetTargetField(),    // maybe will be conflict with record (digest)
		"ttl":         strconv.Itoa(int(rc.TTL)),
	}
```

**Comment by tlimoncelli** [comment]
models.RecordConfig is the struct used internally by dnscontrol.  Each provider includes a function to convert the provider's native format to a models.RecordConfig and the reverse.  

It looks like `requestParams{` is creating the native format for cloudns.  

Maybe if this is a DS record, "record" should be set to rc.DsDigest?

```
req := requestParams{
		"record-type": rc.Type,
		"host":        rc.GetLabel(),
		"ttl":         strconv.Itoa(int(rc.TTL)),
	}
        if rc.Type == "DS" {
            		requestParams["record"] =  rc.DsDigest
                         requestParams["key-tag"]  = rc.DsKeyTag
                         // ...(same for the other fields)...
        } else {
            		requestParams["record"] =  rc.GetTargetField()
        }

```

**Comment by jamaltebi** [comment]
thanks. That what I did  but using case statement.
for each Test( for example create A record) , there are 2 corrections( Modify and delete) ?
I see that it fails in this stage  it founds (`t.FailNow()`) 2 corrections but expect 0 :

/```
/ run a second time and expect zero corrections
		corrections, err = prv.GetDomainCorrections(dom2) // corre ==> invalid length
		if err != nil {
			t.Fatal(err)
		}
		if len(corrections) != 0 { //  found 2 corrections
			t.Logf("Expected 0 corrections on second run, but found %d.", len(corrections))
			for i, c := range corrections {
				t.Logf("#%d: %s", i, c.Msg) // deelte and modify MSg
			}
			t.FailNow() //FailNow marks the function as having failed and stops its execution by calling runtime.Goexit (which then runs all deferred calls in the current goroutine)
		} //
```

![image](https://user-images.githubusercontent.com/51935048/104323927-0d2a8100-54e7-11eb-9ae3-fd3a4986d7ae.png)
even the it fails, DS record can be added

**Comment by tlimoncelli** [comment]
Interesting!

The tests do two runs because there is no other way to test if the changes happened correctly.

The first run puts the desired records in place.  If there is no error, we don't know if the reason is that everything happened correctly or because nothing happened at all.

The second run is exactly like the first run, except there should be zero needed changes.  If changes are needed, this means that the records didn't get inserted properly OR the method of downloading the records is broken.

Can you log into the ClouDNS website and see if the records were created?  Were all the fields correct?

**Comment by jamaltebi** [comment]
After the first test , the DS record is created in clouDNS ( I can log in)
when runing The second , it found 2 corrections
At ende of the test, the NS record and DS record for child( what I added) is deleted

**Comment by tlimoncelli** [comment]
That's good news!  (that the records are created properly).

Usually this means the problem is in `toRc()`.  The records are correct at cloudns but somehow when we download them they don't make it into a RecordConfig properly.

**Comment by jamaltebi** [comment]
maybe because I use **record** (req["record"] = rc.DsDigest)in `toReq `  but in `toRec `(rc.DsDigest = r.DsDigest).
at the first Test zero corrections is running

**Comment by tlimoncelli** [comment]
Yes, it could be the difference between record and DsDigest.

> why the test at the begin delete NS record for child.mydomain and let other NS
> records? I add this record when the test is runing ( because at begin of the test this is deleted)

At the end of each test, the test deletes all records (including NS records). This is because the next test expects an empty zone.  Maybe the NS records have to be deleted after the DS records?

**Comment by jamaltebi** [comment]
yes. but not what happens. for ClouDNS:
-  NS record must be added before DS record
- DS record must be deleted before we can delete NS record 
I want in case Create DS Test,  that no NS record for child must be deleted ?
